File: src/main/java/net/atos/zac/app/admin/converter/RESTZaakbeeindigParameterConverter.java
/*
 * SPDX-FileCopyrightText: 2022 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.app.admin.converter;

import static nl.info.zac.app.zaak.model.RestResultaattypeKt.toRestResultaatType;

import java.util.List;
import java.util.Set;

import jakarta.inject.Inject;

import net.atos.zac.admin.model.ZaakbeeindigParameter;
import net.atos.zac.app.admin.model.RESTZaakbeeindigParameter;
import nl.info.client.zgw.ztc.ZtcClientService;

public class RESTZaakbeeindigParameterConverter {

    @Inject
    private ZtcClientService ztcClientService;

    public List<RESTZaakbeeindigParameter> convertZaakbeeindigParameters(final Set<ZaakbeeindigParameter> zaakbeeindigRedenen) {
        return zaakbeeindigRedenen.stream()
                .map(this::convertZaakbeeindigParameter)
                .toList();
    }

    public static List<ZaakbeeindigParameter> convertRESTZaakbeeindigParameters(
            final List<RESTZaakbeeindigParameter> restZaakbeeindigParameters
    ) {
        return restZaakbeeindigParameters.stream()
                .map(RESTZaakbeeindigParameterConverter::convertRESTZaakbeeindigParameter)
                .toList();
    }

    private RESTZaakbeeindigParameter convertZaakbeeindigParameter(final ZaakbeeindigParameter zaakbeeindigParameter) {
        final RESTZaakbeeindigParameter restZaakbeeindigParameter = new RESTZaakbeeindigParameter();
        restZaakbeeindigParameter.id = zaakbeeindigParameter.getId();
        restZaakbeeindigParameter.zaakbeeindigReden = RESTZaakbeeindigRedenConverter.convertZaakbeeindigReden(zaakbeeindigParameter
                .getZaakbeeindigReden()
        );
        restZaakbeeindigParameter.resultaattype = toRestResultaatType(ztcClientService.readResultaattype(zaakbeeindigParameter
                .getResultaattype()));
        return restZaakbeeindigParameter;
    }

    private static ZaakbeeindigParameter convertRESTZaakbeeindigParameter(final RESTZaakbeeindigParameter restZaakbeeindigParameter) {
        final ZaakbeeindigParameter zaakbeeindigParameter = new ZaakbeeindigParameter();
        zaakbeeindigParameter.setId(restZaakbeeindigParameter.id);
        zaakbeeindigParameter.setZaakbeeindigReden(
                RESTZaakbeeindigRedenConverter.convertRESTZaakbeeindigReden(restZaakbeeindigParameter.zaakbeeindigReden)
        );
        zaakbeeindigParameter.setResultaattype(restZaakbeeindigParameter.resultaattype.getId());
        return zaakbeeindigParameter;
    }
}

File: src/main/java/net/atos/zac/app/admin/converter/RESTZaakbeeindigParameterConverter.kt
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.app.admin.converter

import jakarta.inject.Inject
import net.atos.zac.app.admin.model.RESTZaakbeeindigParameter
import nl.info.client.zgw.ztc.ZtcClientService
import nl.info.zac.admin.model.ZaaktypeCompletionParameters
import nl.info.zac.app.zaak.model.toRestResultaatType

class RESTZaakbeeindigParameterConverter @Inject constructor(
    private val ztcClientService: ZtcClientService
) {
    fun convertZaakbeeindigParameters(
        zaakbeeindigRedenen: Set<ZaaktypeCompletionParameters>
    ): List<RESTZaakbeeindigParameter> =
        zaakbeeindigRedenen.map { convertZaakbeeindigParameter(it) }

    private fun convertZaakbeeindigParameter(
        zaaktypeCompletionParameters: ZaaktypeCompletionParameters
    ): RESTZaakbeeindigParameter = RESTZaakbeeindigParameter().apply {
        id = zaaktypeCompletionParameters.id
        zaakbeeindigReden = RESTZaakbeeindigRedenConverter.convertZaakbeeindigReden(
            zaaktypeCompletionParameters.zaakbeeindigReden
        )
        resultaattype = ztcClientService.readResultaattype(
            zaaktypeCompletionParameters.resultaattype
        ).toRestResultaatType()
    }
}

        fun convertRESTZaakbeeindigParameters(
            restZaakbeeindigParameters: List<RESTZaakbeeindigParameter>
        ): List<ZaaktypeCompletionParameters> =
            restZaakbeeindigParameters.map { convertRESTZaakbeeindigParameter(it) }

        private fun convertRESTZaakbeeindigParameter(
            restZaakbeeindigParameter: RESTZaakbeeindigParameter
        ): ZaaktypeCompletionParameters = ZaaktypeCompletionParameters().apply {
            id = restZaakbeeindigParameter.id
            zaakbeeindigReden = RESTZaakbeeindigRedenConverter.convertRESTZaakbeeindigReden(
                restZaakbeeindigParameter.zaakbeeindigReden
            )
            resultaattype = restZaakbeeindigParameter.resultaattype.id
        }

